flips/20220817-extensibility.md

```diff
@@ -0,0 +1,423 @@
+a declaration kind. Any fields and methods on the original type would be accessible to the extension on the `self` value. 
+Note, however, that extensions only have access to the same fields and functions as other code declared in the same place would have; 
+i.e. an extension defined in the same contract as its original type would have access to `pub` and `access(contract)` fields and methods, but not `priv` fields
+or methods, while an extension defined in a different contract and account to its original type would only be able to reference `pub` fields and methods on its 
+original type. Extensions may also not override or overload fields or methods on the original type, even if the original fields are not accessible to the extension. 
+Hence, if the original type `T` declared a `priv` field named `foo`, an extension `E` of `T` would not be able to declare any fields or methods named `foo`, even
+though `foo` is not accessible to `E`. This is to avoid any confusion or ambiguity when reading code. 
+
+Interface conformances are checked against the combination of the original type with the extension applied. In particular, this means that it is possible to declare
+an extension with an interface annotation if neither the extension nor the original type conform to that interface, as long as the resultant extended type would conform. 
+For example:
+
+```cadence
+pub struct interface I {
+    pub fun foo(): Int
+    pub fun bar(): String
+}
+
+pub struct S {
+    pub fun foo(): Int { ... }
+}
+
+pub extension E for S: I {
+    pub fun bar(): String { ... }
+}
+```
+
+In this case, `E`'s declaration conforms to `I` because type of `S` with the extension `E` applied produces a type that conforms to `I`. 
+
+Any additional fields that are declared in an extension must be initialized, just as any fields declared in a composite must be. An extension
+that declares fields must declare an additional (partial) initializer, which is run when the extension is created. For this reason, the `init`
+method in an extension cannot make reference to any fields or methods of the base type, as it will be run before extension is attached to that type. It is
+recommended that users simply initialize any fields as simply as possible here, and save any complex set-up for the extension's `attach` method (explained
+later in this section).
+
+The same checks on normal initializers apply to extension initializers; namely that all the fields declared in the extensions must receive a value in the
+extension's initializer. So, the following would be a legal extension:
+
+```cadence
+pub struct S {}
+
+pub extension E for S {
+    pub let x: String
+    init(_ x: String) {
+        self.x = x
+    }
+}
+```
+
+while this would not:
+
+```cadence
+pub struct S {}
+
+pub extension E for S {
+    pub let x: String
+    pub let y: String
+    init(_ x: String) {
+        self.x = x
+    }
+}
+```
+
```

Comment:
Maybe also add an example where `S` has a field `z` and `E` tries to use it in its initializer, which is invalid.